presto-hive/src/main/java/com/facebook/presto/hive/SortingFileWriter.java

```diff
@@ -0,0 +1,207 @@
+
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_TOO_MANY_BUCKET_SORT_FILES;
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_WRITER_CLOSE_ERROR;
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_WRITER_DATA_ERROR;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static java.util.Objects.requireNonNull;
+
+public class SortingFileWriter
+        implements HiveFileWriter
+{
+    private static final Logger log = Logger.get(SortingFileWriter.class);
+
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(SortingFileWriter.class).instanceSize();
+
+    private final FileSystem fileSystem;
+    private final Path tempFilePrefix;
+    private final int maxTempFiles;
+    private final List<Type> types;
+    private final List<Integer> sortFields;
+    private final List<SortOrder> sortOrders;
+    private final HiveFileWriter outputWriter;
+    private final SortBuffer sortBuffer;
+    private final List<Path> tempFiles = new ArrayList<>();
+
+    public SortingFileWriter(
+            FileSystem fileSystem,
+            Path tempFilePrefix,
+            HiveFileWriter outputWriter,
+            DataSize maxMemory,
+            int maxTempFiles,
+            List<Type> types,
+            List<Integer> sortFields,
+            List<SortOrder> sortOrders,
+            PageSorter pageSorter)
+    {
+        checkArgument(maxTempFiles > 0, "maxTempFiles must be greater than zero");
+        this.fileSystem = requireNonNull(fileSystem, "fileSystem is null");
+        this.tempFilePrefix = requireNonNull(tempFilePrefix, "tempFilePrefix is null");
+        this.maxTempFiles = maxTempFiles;
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.sortFields = ImmutableList.copyOf(requireNonNull(sortFields, "sortFields is null"));
+        this.sortOrders = ImmutableList.copyOf(requireNonNull(sortOrders, "sortOrders is null"));
+        this.outputWriter = requireNonNull(outputWriter, "outputWriter is null");
+        this.sortBuffer = new SortBuffer(maxMemory, types, sortFields, sortOrders, pageSorter);
+    }
+
+    @Override
+    public long getWrittenBytes()
+    {
+        return outputWriter.getWrittenBytes();
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return INSTANCE_SIZE + sortBuffer.getRetainedBytes();
+    }
+
+    @Override
+    public void appendRows(Page page)
+    {
+        if (tempFiles.size() == maxTempFiles) {
+            throw new PrestoException(HIVE_TOO_MANY_BUCKET_SORT_FILES, "Too many temporary files for sorted bucket writer");
+        }
+
+        if (!sortBuffer.canAdd(page)) {
+            flushBuffer();
+        }
+        sortBuffer.add(page);
+    }
+
+    @Override
+    public void commit()
+    {
+        if (!sortBuffer.isEmpty()) {
```

Comment:
If there are no temp files, then just "flushTo(outputWriter::appendRows)", close and return.